saleor/graphql/order/mutations/fulfillments.py

```diff
@@ -119,18 +141,55 @@ def clean_lines(cls, order_lines, quantities):
                     }
                 )
 
+    @classmethod
+    def check_warehouses_for_duplicates(cls, warehouse_ids):
+        for warehouse_ids_for_line in warehouse_ids:
+            duplicates = get_duplicated_values(warehouse_ids_for_line)
+            if duplicates:
+                raise ValidationError(
+                    {
+                        "warehouse": ValidationError(
+                            "Duplicated warehouse ID.",
+                            code=OrderErrorCode.DUPLICATED_INPUT_ITEM,
+                        )
+                    }
+                )
+
+    @classmethod
+    def check_lines_for_duplicates(cls, lines_ids):
+        duplicates = get_duplicated_values(lines_ids)
+        if duplicates:
+            raise ValidationError(
+                {
+                    "orderLineId": ValidationError(
+                        "Duplicated order line ID.",
+                        code=OrderErrorCode.DUPLICATED_INPUT_ITEM,
+                    )
+                }
+            )
+
     @classmethod
     def clean_input(cls, data):
         lines = data["lines"]
-        quantities = [line["quantity"] for line in lines]
+
+        warehouse_ids_for_lines = [
+            [stock["warehouse"] for stock in line["stocks"]] for line in lines
+        ]
+        cls.check_warehouses_for_duplicates(warehouse_ids_for_lines)
+
+        quantities_for_lines = [
+            [stock["quantity"] for stock in line["stocks"]] for line in lines
+        ]
+
         lines_ids = [line["order_line_id"] for line in lines]
+        cls.check_lines_for_duplicates(lines_ids)
         order_lines = cls.get_nodes_or_error(
             lines_ids, field="lines", only_type=OrderLine
         )
 
-        cls.clean_lines(order_lines, quantities)
+        cls.clean_lines(order_lines, quantities_for_lines)
 
-        if sum(quantities) <= 0:
+        if sum(sum(quantities_for_lines, [])) <= 0:
```

Comment:
sum within a sum with 1st arg from double list comprehension loop and 2nd sum argument of array.
maybe it's just me but this is super confusing at 1st glance, perhaps body of 1st sum could be extracted to meaningfull var or even entire thing to a separate func?

Reply:
Agree